homeassistant/components/weatherflow_cloud/sensor.py

```diff
@@ -176,35 +302,161 @@ async def async_setup_entry(
     )
 
+    entities.extend(
+        WeatherFlowWebsocketSensorObservation(
+            coordinator=observation_coordinator,
+            description=sensor_description,
+            station_id=station_id,
+            device_id=device_id,
+        )
+        for station_id in observation_coordinator.stations.station_outdoor_device_map
+        for device_id in observation_coordinator.stations.station_outdoor_device_map[
+            station_id
+        ]
+        for sensor_description in WEBSOCKET_OBSERVATION_SENSORS
+    )
+    async_add_entities(entities)
 
-class WeatherFlowCloudSensor(WeatherFlowCloudEntity, SensorEntity):
-    """Implementation of a WeatherFlow sensor."""
 
-    entity_description: WeatherFlowCloudSensorEntityDescription
+class WeatherFlowSensorBase(WeatherFlowCloudEntity, SensorEntity, ABC):
+    """Common base class."""
 
     def __init__(
         self,
-        coordinator: WeatherFlowCloudDataUpdateCoordinator,
-        description: WeatherFlowCloudSensorEntityDescription,
+        coordinator: WeatherFlowCloudUpdateCoordinatorREST
+        | WeatherFlowCloudDataCallbackCoordinator[
+            EventDataRapidWind | WebsocketObservation,
+            RapidWindListenStartMessage | ListenStartMessage,
+            RapidWindWS | WebsocketObservation,
+        ],
+        description: WeatherFlowCloudSensorEntityDescription
+        | WeatherFlowCloudSensorEntityDescriptionWebsocketWind
+        | WeatherFlowCloudSensorEntityDescriptionWebsocketObservation,
         station_id: int,
+        device_id: int | None = None,
     ) -> None:
-        """Initialize the sensor."""
-        # Initialize the Entity Class
+        """Initialize a sensor."""
         super().__init__(coordinator, station_id)
+        self.station_id = station_id
+        self.device_id = device_id
         self.entity_description = description
-        self._attr_unique_id = f"{station_id}_{description.key}"
+        self._attr_unique_id = self._generate_unique_id()
+
+    def _generate_unique_id(self) -> str:
+        """Generate a unique ID for the sensor."""
+        if self.device_id is not None:
+            return f"{self.station_id}_{self.device_id}_{self.entity_description.key}"
+        return f"{self.station_id}_{self.entity_description.key}"
+
+    @property
+    @abstractmethod
+    def native_value(self) -> StateType | date | datetime | Decimal:
+        """Abstract method for native value."""
+
+    @property
+    def available(self) -> bool:
+        """Get if available."""
+
+        if not super().available:
+            return False
+
+        if self.device_id is not None:
+            # Websocket sensors - have Device IDs
+            return bool(
+                self.coordinator.data
+                and self.coordinator.data[self.station_id][self.device_id] is not None
+            )
+
+        # Check availability of a REST based sensor
+        return bool(self.coordinator.data)
```

Comment:
why do we bool this? If there is no coordinator data, the update likely failed and then super().available wouldve caught it

Reply:
I changed this to `return True` and you are right it seems to be fine. I'll commit it.